components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -1437,8 +1417,59 @@ void ConversationHandler::OnEngineCompletionComplete(
         SetAPIError(mojom::APIError::None);
       }
     }
+    CompleteGeneration(false);
+    return;
   }
 
+  // Handle success, which might mean do nothing much since all data was passed
+  // in the streaming "received" callback.
+  DVLOG(2) << __func__ << ": With value";
+  if (result->event && result->event->is_completion_event() &&
+      !result->event->get_completion_event()->completion.empty()) {
+    UpdateOrCreateLastAssistantEntry(std::move(*result));
+  } else {
+    // This is a workaround for any occasions where the engine returns
+    // a success but there was no new entry.
+    if (needs_new_entry_) {
+      SetAPIError(mojom::APIError::ConnectionIssue);
+      CompleteGeneration(false);
+      return;
+    }
+  }
+  OnConversationEntryAdded(chat_history_.back());
+
+  // Check if we need title generation (after assistant response is added)
+  if (engine_->RequiresClientSideTitleGeneration() &&
+      chat_history_.size() == 2) {
+    // Keep the request active and complete the generation in OnTitleGenerated.
+    engine_->GenerateConversationTitle(
+        associated_content_manager_->GetCachedContentsMap(), chat_history_,
+        base::BindOnce(&ConversationHandler::OnTitleGenerated,
+                       weak_ptr_factory_.GetWeakPtr()));
+    return;
+  }
```

Comment:
I have to introduce this to conversation handler because it needs the first assistant response for uploads (image, PDF, ...), and we process streaming events and update histories in conversation handler.